core/services/job/orm.go

```diff
@@ -846,21 +846,35 @@ func (o *orm) PipelineRunsByJobsIDs(ids []int32) (runs []pipeline.Run, err error
 // FindPipelineRunIDsByJobID fetches the ids of pipeline runs for a job.
 func (o *orm) FindPipelineRunIDsByJobID(jobID int32, offset, limit int) (ids []int64, err error) {
 	err = o.q.Transaction(func(tx pg.Queryer) error {
+		var maxId int64
+		if err = tx.Select(&maxId, "SELECT MAX(id) FROM pipeline_runs"); err != nil {
+			return errors.Wrap(err, "error while loading runs")
+		}
+
 		stmt := `
-SELECT pipeline_runs.id
-FROM pipeline_runs
-WHERE pipeline_runs.pipeline_spec_id = (SELECT jobs.pipeline_spec_id FROM JOBS WHERE jobs.id = $1)
-ORDER BY pipeline_runs.created_at DESC, pipeline_runs.id DESC
-OFFSET $2
-LIMIT $3
-`
-		if err = tx.Select(&ids, stmt, jobID, offset, limit); err != nil {
-			return errors.Wrap(err, "error loading runs")
+				SELECT pipeline_runs.id
+				FROM pipeline_runs
+				WHERE pipeline_runs.pipeline_spec_id = (SELECT jobs.pipeline_spec_id 
+														FROM JOBS WHERE jobs.id = $1)
+				AND id >= $4 AND id <= $5 FROM pipeline_runs) - $4
+				ORDER BY pipeline_runs.created_at DESC, pipeline_runs.id DESC
+				OFFSET $2
+				LIMIT $3
+			`
+		// Only search the most recent n jobs (whether deleted or not), starting with n = 1000 and
+		//  doubling only if we still need more.  Without this, large tables can result in the UI
+		//  becoming unusably slow, continuously flashing, or timing out.  The ORDER BY in
+		//  this query requires a sort of all runs matching jobID, so we restrict it to the
+		//  range minId <-> maxId.
+		for n := int64(1000); n < maxId; n *= 2 {
```

Comment:
what if minId = 0 and maxId < 1000? this loop will never execute?

Reply:
I think this is a good call here... We need to make sure this returns even if the amount of runs in the database are less than 1000 I believe.

cc: @jmank88

Reply:
Good point, I need to refactor this so that the n < maxId check comes at the end of the for block instead of the beginning.